lib/spack/spack/package.py

```diff
@@ -2195,6 +2195,85 @@ def possible_dependencies(*pkg_or_spec, **kwargs):
     return visited
 
 
+#: Registers which are the detectable packages, by repo and package name
+detectable_packages = collections.defaultdict(list)
+
+
+def detectable(decorated_cls):
+    """Mark a package class as detectable and add a default implementation
+    for detection logic.
+
+    The decorator registers the class as detectable and attaches to it a
+    default implementation of ``determine_spec_details``.
+
+    Args:
+        decorated_cls: class to be decorated
+
+    Returns:
+        Decorated class
+    """
+    @classmethod
+    def determine_spec_details(cls, prefix, exes_in_prefix):
+        """Allow ``spack external find ...`` to locate installations.
+
+        Args:
+            prefix (str): the directory containing the executables
+            exes_in_prefix (set): the executables that match the regex
+
+        Returns:
+            The list of detected specs for this package
+        """
+        exes_by_version = collections.defaultdict(list)
+        # The default filter function is the identity function for the
+        # list of executables
+        filter_fn = getattr(cls, 'filter_detected_exes', lambda x, exes: exes)
+        exes_in_prefix = filter_fn(prefix, exes_in_prefix)
+        for exe in exes_in_prefix:
+            try:
+                version_str = cls.determine_version(exe)
+                if version_str:
+                    exes_by_version[version_str].append(exe)
```

Comment:
Why append? If I have a prefix with `python`, `python2`, `python2.7`, `python3`, and `python3.7` executables, I would like Spack to find exactly 2 external package installations (2.7 and 3.7). I guess you could theoretically have multiple copies of GCC at the same version in the same prefix...

Reply:
Actually, I think I understand this now. This is what you were saying about needing both `cmake` and `ccmake` exes in order to determine the spec. So later this would call `cls.determine_variants` once for each version (2.7 and 3.7), and although I may get multiple exes, I only need to handle one. If so, I'm fine with this implementation.

Reply:
> So later this would call cls.determine_variants once for each version (2.7 and 3.7), and although I may get multiple exes, I only need to handle one.

Exactly that :+1: